### DISCUSSION TITLE
Ensuring Proper Initialization Order for Beans in Micronaut

### DISCUSSION THREAD

**Discussion by rsalbergaria** [micronaut-projects/micronaut-core#11282] [Q&A]
Hi Micronaut Team,

I am currently working on a project where I need to provide KafkaPublisher beans programmatically. The goal is to ensure that these beans are initialized correctly and in the right order, especially when developers use named instances.

Here is a brief overview of my setup:

1. I have a factory class that provides KafkaPublisher beans.
2. Developers can instantiate KafkaPublisher without a name, which should return a default instance.
3. Developers can also use Named to request specific instances created programmatically.

I also have a BeanCreatedEventListener to capture the creation of KafkaPublisher beans:

> `@Singleton
public class KafkaPublisherBeanListener implements BeanCreatedEventListener<KafkaPublisher> {

    @Override
    public KafkaPublisher onCreated(BeanCreatedEvent<KafkaPublisher> event) {
        String beanName = event.getBeanDefinition().getName();
        System.out.println("Bean created with name: " + beanName);

        KafkaPublisher kafkaPublisher = event.getBean();
        if ("defaultKafkaPublisher".equals(beanName)) {
            // Configure the default KafkaPublisher
        } else if ("customKafkaPublisher".equals(beanName)) {
            // Configure the custom KafkaPublisher
        }

        return kafkaPublisher;
    }
}`

The issue I am facing is that not all executions are passing through the BeanCreatedEventListener. I need to ensure that my factory instances are initialized first and that the correct KafkaPublisher instance is returned based on the @Named annotation.

Could you please provide guidance on how to ensure the proper initialization order and correct instance creation in Micronaut? Any help or suggestions would be greatly appreciated.

**Comment by rsalbergaria** [comment]
Can you help me? @dstepanov

**Comment by rsalbergaria** [comment]
Can this issue help me? https://github.com/micronaut-projects/micronaut-core/issues/11044

**Comment by dstepanov** [comment]
What executions are not passed through BeanCreatedEventListener? Can you create a simple project?

**Reply by rsalbergaria** [reply]
Certainly, here’s the repository: [GitHub - demo-aop](https://github.com/rsalbergaria/demo-aop). I'm not sure if I’m on the right path for what I want to achieve, so any advice would be appreciated.

The goal is to instantiate KafkaPublisher beans programmatically in my library based on a configuration file, allowing other developers to inject these beans using @Named. Perhaps there’s a better approach to creating these beans?

My initial idea was to intercept every attempt to create a KafkaPublisher, receive a context of the @Named identifier used, and then return the correct bean.
